Add CAPQ, upgrade SprayList, upgrade Linden and add new benchmark

kjellwinblad opened the pull request:
This pull request contains three commits:

1. The first commit adds the priority queues CA-PQ, CA-DM, CA-IN and CATree to the random benchmark. They are all described in the LCPC'2016 paper [The Contention Avoiding Concurrent Priority Queue][1].
2. The second commit upgrades both the SprayList implementation and the Linden priority queue implementation.
3. The last commit adds a new shortest paths benchmark that reads the input graph from a file.

Please, expand the commit messages below for details about these changes.

The benchmark code and the priority queues have been tested by running the new shortest paths with up to 64 worker threads on a machine with 64 hardware threads. To check that the priority queues and the benchmark work as intended, the calculated distances were compared against the actual distances after each run. 

[1]: https://www.it.uu.se/research/group/languages/software/ca_pq

kjellwinblad committed:
Add implementation of the contention avoiding priority queue

The Contention Avoiding Priority Queue (CA-PQ) is a relaxed priority
queue described in the paper "The Contention Avoiding Concurrent
Priority Queue" which is published in the 29th International Workshop
on Languages and Compilers for Parallel Computing (LCPC 2016). A
preprint version of the paper can be found at
https://www.it.uu.se/research/group/languages/software/ca_pq . This
commit adds CA-PQ as well as the priority queues CA-DM, CA-IN and
CATree to the random benchmark. CA-DM, CA-IN and CATree are derived
from CA-PQ by turning off features and are all described in the paper
mentioned above.

kjellwinblad committed:
Update the SprayList and the Linden priority queue

The code for the SprayList and the Linden priority queue is updated to
correspond to a patched version of the latest code from the SprayList
repository (*). In contrast to the previous version of the Linden
priority queue implementation, the version included in this commit
supports duplicated keys. In contrast to the previous version of the
SprayList implementation, the version included in this commit supports
associating a value with the keys.

The patch that has been applied on top of the latest SprayList version
(*) improves SprayList's memory management. (A pull request will soon
be submitted to the SprayList repository with this patch.)

Without the patch it is only possible to specify how much memory each
thread will be able to allocate. This is problematic for the single
source shortest paths benchmark as it makes it impossible to run the
benchmark with a single thread and many threads using the same
binary. The benchmark would either run out of memory when a single
thread is used or there would not be enough memory on the machine when
many threads are used.

With the patch one can specify how much memory all threads can
allocated together (the #define SSALLOC_SIZE_ALL in
include/ssalloc.h). The patch makes sure that this amount of memory is
divided between the threads (Note that the global variable
number_of_threads has to be set to the correct value for this to
work).

(*) https://github.com/jkopinsky/SprayList/tree/e6c851d93156646efa4ce0a9dbdb2e1a3e8d3693

schuay commented on bench.py:
```diff
@@ -5,7 +5,11 @@
 from optparse import OptionParser
 
-ALGORITHMS = [ 'cheap'
+ALGORITHMS = [ 'capq'
```
Nit: Please alpha-sort these.

kjellwinblad replied:
Will fix.

schuay commented on lib/capq/binheap_test.c:
```diff
@@ -0,0 +1,40 @@
+#include <stdio.h>
+#include "binheap.h"
+
+int main()
```
Nit: Could add this to the test suite (or remove it).

kjellwinblad replied:
I don't think it hurts to keep it where it is but I can remove it or move it if you insist?

schuay replied:
It's fine, up to you. I don't care too much about stuff in lib/ :)

schuay commented on src/bench/random.cpp:
```diff
@@ -31,6 +31,7 @@
 #include <valgrind/callgrind.h>
 #endif
 
+#include "pqs/cppcapq.h"
```
Nit: Please alpha-sort pqs/.

kjellwinblad replied:
Will fix.

schuay commented on src/bench/random.cpp:
```diff
@@ -188,15 +199,17 @@ usage()
-            PQ_LINDEN, PQ_LSM, PQ_MLSM, PQ_MULTIQ, PQ_SEQUENCE, PQ_SKIP, PQ_SLSM, PQ_SPRAY);
+            PQ_CAPQ, PQ_CADM, PQ_CAIN, PQ_CATREE, PQ_CHEAP, PQ_DLSM, PQ_GLOBALLOCK, PQ_KLSM16,
+            PQ_KLSM128, PQ_KLSM256, PQ_KLSM4096, PQ_LINDEN, PQ_LSM, PQ_MLSM, PQ_MULTIQ, PQ_SEQUENCE,
+            PQ_SKIP, PQ_SLSM, PQ_SPRAY, "nada");
```
What's the point of the last "nada" entry?

kjellwinblad replied:
My mistake. I will remove it.

schuay commented on src/bench/random.cpp:
```diff
@@ -378,12 +391,13 @@ bench_thread(PriorityQueue *pq,
                              settings.size - thread_id * slice_size : slice_size;
     for (int i = 0; i < initial_size; i++) {
         uint32_t elem = keygen.next();
+        uint32_t value = keygen.next();
```
Is it beneficial to add a separate value? I reused elem since the value seems irrelevant and it just adds keygen overhead. Same below.

kjellwinblad replied:
My mistake again. I think I added a separate value to test something and then I must have forgotten to change it back.

schuay commented on src/bench/pqs/linden.cpp:
```diff
@@ -35,16 +36,43 @@ Linden::~Linden()
+{   
+    unsigned long k_ret;
+    v = (uint32_t)deletemin_key(m_q->pq, &k_ret);
+    if(k_ret == ((unsigned long)-1)){
```
Nit: return (k_ret != -1)

kjellwinblad replied:
Will fix.

schuay commented on src/bench/pqs/linden.cpp:
```diff
@@ -35,16 +36,43 @@ Linden::~Linden()
+    unsigned long k_ret;
+    v = deletemin_key(m_q->pq, &k_ret);
+    k = k_ret -1;
+    if(k_ret == ((unsigned long)-1)){
```
Same here.

kjellwinblad replied:
Will fix.

schuay commented on lib/spraylist_linden/ssalloc.c:
```diff
@@ -30,14 +30,20 @@ ssalloc_set(void* mem)
   for (i = 0; i < SSALLOC_NUM_ALLOCATORS; i++)
     {
+#ifdef SSALLOC_SIZE_ALL
+      ssalloc_app_mem[i] = (void*) memalign(64, SSALLOC_SIZE_ALL / number_of_threads);
```
Just wondering, why can't we pass this into ssalloc_init(int number_of_threads) instead of using a global variable?

kjellwinblad replied:
I don't think there is much reason not to do that. I will try to remove the global variable.

schuay commented on src/k_lsm/k_lsm_inl.h:
```diff
@@ -82,6 +82,55 @@ k_lsm<K, V, Rlx>::insert(const K &key,
 
 template <class K, class V, int Rlx>
 bool
+k_lsm<K, V, Rlx>::delete_min(K &key, V &val)
```
Please don't duplicate all of this code. Why not implement delete_min(val) as 

delete_min(V &val) {
  K key;
  return delete_min(key, val);
}

Same in all other files of this commit.

schuay replied:
Also, can you split the delete_min changes into a separate CL?

kjellwinblad replied:
I agree. There is no reason to duplicate all that code. Will fix that and split out the delete_min changes to a separate commit.

schuay reviewed (commented):
Awesome! Looks good to me in general, with a few minor comments inline.

kjellwinblad committed:
Add delete_min method that returns both key and value

This commit adds a delete_min method that returns both the key and the
value to the priority queues that don't already have such a method. It
is useful to get both the key and the value from a deleted item in
algorithms such as Dijkstra's single source shortest paths algorithm.

kjellwinblad commented:
Thank you for the feedback! I will update the pull request as soon as I have got enough time to do the changes.

kjellwinblad committed:
Add shortest paths benchmark that load graph from file

This shortest path benchmark (file_shortest_paths.cpp) is implemented
as a new benchmark instead of as a new option in the already existing
shortest paths benchmark (shortest_paths.cpp) that generates a random
graph instead of loading a graph from a file. The reason for this is
that the new benchmark is a heavily modified version of the old
version. Task allocation and the global counter num_tasks is avoided
in the new benchmark. Furthermore the verify_graph function is not
present as the calculated paths can be verified by verifying the
output file. The implementation of this benchmark is inspired by the
sssp benchmark in the SprayList repository (See
lib/spraylist_linden/sssp.c).

Additionally this commit contains a program that can be used to
generate input graphs for the benchmark (generate_random_graph.cpp).

kjellwinblad commented:
I have made changes according to your feedback. Please, check that everything looks alright and let me know if you want me to fix something more.

schuay commented on src/bench/pqs/multiq_inl.h:
```diff
@@ -36,13 +36,13 @@ multiq<K, V, C>::~multiq()
 template <class K, class V, int C>
 bool
-multiq<K, V, C>::delete_min(V &value)
+    multiq<K, V, C>::delete_min(K &key, V &value)
```
Nit: Whitespace.

schuay commented on src/bench/pqs/multiq_inl.h:
```diff
@@ -57,11 +57,17 @@ multiq<K, V, C>::delete_min(V &value)
+            // TODO: Not a permanent solution
             unlock(i);
-            return false;
+            if(retry_count == 0){
```
Nit: Please preserve the whitespace style of the remaining code.

schuay commented on src/bench/pqs/multiq_inl.h:
```diff
@@ -57,11 +57,17 @@ multiq<K, V, C>::delete_min(V &value)
+                return false;
+            }else{
+                retry_count--;
+                goto delete_min_start_label;
```
We can remove the label and just do a 'continue'.

schuay commented on src/bench/pqs/linden.cpp:
```diff
@@ -35,16 +36,38 @@ Linden::~Linden()
+}
+
+void
+Linden::insert(const size_t &key,
```
Nit: Is this overload really needed? Same for all other sites. Can't we instead make sure that call-sites all use uint32_t's?

kjellwinblad replied:
I think we could make sure that call sites use uint32_t but that would require casting from size_t or other changes so I don't think it is a good idea. Do you have any particular reason for wanting to avoid the overloads?

schuay replied:
I just find it unnecessary code duplication. Wouldn't it be an option to settle on 32- or 64-bit ints for use in benchmarks? Another idea would be to use templated insert() and delete_min() methods?

kjellwinblad replied:
I think it is an excellent idea to settle on 64-bit keys and values. I think 64-bit would be better than 32-bit as:

* that is what the priority queues that are implemented in C (Linden, SprayList and CA-PQ) are hard coded for on modern platforms and
* it would make it possible to fit a memory address in the value field.

However, I think such a change is not related to this pull request so perhaps it is better to do this as a separate commit once this pull request has been accepted. What do you think?

schuay replied:
Ok, let's do that.

schuay commented on src/bench/pqs/spraylist.cpp:
```diff
@@ -55,14 +55,38 @@ void
+    unsigned long v_ret;
+    int ret;
+    do {
+        ret = spray_delete_min_key(m_q, &k_ret, &v_ret, d);
```
Why did this change? I couldn't find any documentation about this in the spraylist code above. Please add a comment explaining the condition below, assuming it's the right thing to do.

kjellwinblad replied:
It requires some detective work to figure out that this is how one is supposed to use spray_delete_min_key. I figured it out from the "sssp.c" benchmark in the SprayList repository *. My understanding is that spray_delete_min_key can sometimes fail even though the SprayList is not empty, in which case spray_delete_min_key returns 0 and sets the key to a value that is different from -1. I agree that we should document why we are using spray_delete_min_key in this way. Can you first check that you agree that what I did seems correct?

@jkopinsky , can you also check that spray_delete_min_key is used in the correct way here?

* https://github.com/kjellwinblad/klsm/blob/1ca3340a752dcf57400a9a9b67e84b2ae8bacf94/lib/spraylist_linden/sssp.c#L158

schuay replied:
I see, thanks for the explanation. I'm not sure it's a good idea to loop within delete_min though. For instance, the klsm can also fail spuriously but that is part of the contract for delete_min.

I think handling these cases (potentially with loops) should be done outside delete_min. Not sure how the SprayList works internally though, so I may be wrong.

kjellwinblad replied:
I did it this way because it makes the file_shortest_paths benchmark (src/bench/file_shortest_paths.cpp) together with the SprayList behave in the same way as the shortest paths benchmark implemented by @jkopinsky (lib/spraylist_linden/sssp.c). I don't see any other good way to accomplish this but I'm open for suggestions.

If I'm not mistaken, the file_shortest_paths benchmark used with the SprayList could compute the wrong result if this loop was not present even though it is probably very unlikely in practice. The reason is that the spurious failures seems to depend on a random number generator so if all worker threads spuriously fail enough times at the wrong time the algorithm will stop even though there is still work to do. Again if I'm not mistaken, the file_shortest_paths benchmark together with KLSM will always compute the correct result even though KLSM can spuriously fail. The reason is that all worker threads can't fail "at the same time" with KSLM. The stop condition for the file_shortest_paths benchmark is that all worker threads called delete_min after some time point after which no insert operations have been issued and got false as return value.

jkopinsky replied:
Yes, currently SprayList can fail spuriously either due to failed Compare-and-Swap (unlikely by design) or due to randomness even if the list is non-empty (also unlikely, though chance of this increases if the list is very small). 

I think it was a planned feature to have retries happen internally but this never got implemented; sorry for the confusion.

schuay replied:
The src/bench/shortest_paths.cpp benchmark has a stop condition based on a global counter of the number of active tasks, so I think it calculates the correct result regardless of spurious failures - is that not the case? See https://github.com/klsmpq/klsm/blob/master/src/bench/shortest_paths.cpp#L192, a failure just results in retrying the deletion.

Anyway, if a non-spurious failure mode does turn out to be required, maybe we should add it as another method to avoid influencing existing benchmarks.

Once this question is resolved, it's all good to land from my side :+1:

kjellwinblad replied:
@schuay Yes, I think you are right, src/bench/shortest_paths.cpp calculates the correct result regardless of spurious failures. Unfortunately, the global counter is a serious scalability problem. That is why src/bench/file_shortest_paths.cpp does not depend on a global counter. Instead src/bench/file_shortest_paths.cpp use a stop condition that grantee correctness as long as the priority queue has the property described in Theorem 4  in the paper  [The Contention Avoiding Concurrent Priority Queue LCPC'2016](http://www.it.uu.se/research/group/languages/software/ca_pq/relax_prio_queue_paper.pdf). I'm quite sure that KLSM satisfies this property and I don't see any reason why SprayList shouldn't.

I was surprised to see that there seems to be a small difference between using the retry loop and not using the retry loop as retries should be rare (see measurements below). Anyway, my opinion is that the proper version to measure is the one with the retry loop. I interpret @jkopinsky's answer above as that this is also his opinion. The only reason I see to expose SprayList's spurious failures to the application is to allow the application to add a custom yield or sleep before the retry. Changing "delete_min(uint32_t &v)" to look as the code that can be found below will avoid influencing the existing benchmarks. Do you want me to do that?

## SprayList Measurements
The measurements where done using the version with commit id 1ca3340a752dcf57400a9a9b67e84b2ae8bacf94. I ran it on a "AMD FX(tm)-6100 Six-Core Processor". The "retry loop version" has the retry loop and the "without retry loop" version has a delete_min method modified to look like this:

```
bool
spraylist::delete_min(uint32_t &v)
{
    size_t k_ret;
    size_t v_ret;
    const int ret = spray_delete_min_key(m_q, &k_ret, &v_ret, d);
    v = (uint32_t)v_ret;
    return ret == 1;
} 
```


### retry loop version 
```
./build/src/bench/random -p 6 spray
3890564
$ ./build/src/bench/random -p 6 spray
3893154
$ ./build/src/bench/random -p 6 spray
3902697
```
### without retry loop
```
$ ./build/src/bench/random -p 6 spray
4049361
$ ./build/src/bench/random -p 6 spray
4063919
$ ./build/src/bench/random -p 6 spray
4049860
```

schuay replied:
Wow that's a pretty noticeable difference. Anyway, I'll defer to your best judgement on this since it's  been a while that I looked at all of this.

My gut reaction is that 1. it'd be nice to keep the behavior of existing benchmarks consistent (unless they're incorrect), 2. behavior of different PQs should be consistent (it seems weird if only some loop in delete_min), and 3. making the shortest path benchmark more scalable sounds good.

Let me know when you think this is ready to merge.

kjellwinblad replied:
I will try to give answers to the first two of your gut reactions (please let me know if you don't agree with something):

1. I'm not sure if the old measurements in the random benchmark with the spraylist are incorrect given that you also measure the quality of the delete_min results. No retry loop in delete_min should give better performance but worse quality and the other way around. However, I think the chance that the results have relevance for a real application is larger with the retry loop as that is what one would like to have in a real application.
2. I think the reason you think it is weird to only have the retry loop for SparyList is that the retry loop should actually be inside SprayList's "own" delete_min function. I bet KLSM has some retry loop internally that instead could be handled by spurious failures, right? The spurious failures that the retry loop retries on are quite different from KLSM's spurious failures (I think) in how easily detectable they are.

schuay replied:
Re 2.: The KLSM doesn't have an internal retry loop, it only retries once after spying if the first attempt failed IIRC. It just seems weird to me that some delete_min's have a retry loop and others don't.

kjellwinblad replied:
The KLSM as decribed in the [The Lock-free k-LSM Relaxed Priority Queue](https://arxiv.org/abs/1503.05698) paper does have a retry loop (See lines 15-25 in Listing 5). Doesn't the actual implementation have the corresponding loop? I totally agree that it is weird to have it for the SprayList and not for the KLSM, if the KLSM doesn't have this loop.

schuay replied:
You're right, there is a retry loop here: https://github.com/klsmpq/klsm/blob/master/src/k_lsm/k_lsm_inl.h#L104

It retries if both the local dlsm and the shared slsm are empty, and spying (taking items from another thread's local dlsm) copied more than 0 items to the local dlsm.

It differs from Listing 5 in that failed CAS result in a spurious failure.

I'm fine with keeping the SprayList loop if you think that is appropriate (it'd be great to have a short comment as to what the conditions (return value, key value) mean).

kjellwinblad replied:
> It differs from Listing 5 in that failed CAS result in a spurious failure.

Why doesn't the implementation do as the paper? I think that it would make more sense if the implementation retried on CAS failure as the code in the paper does. I think so for the same reasons as I think the retry loop that I added to SprayList's delete_min should be there.

I was wrong in thinking that the KLSM implementation didn't have this kind of spurious failures. I thought the KLSM implementation's spurious failures happened when the issuing thread's local queue and the shared queue both were empty but some other local queue had items, which I could understand as it would be costly to detect this kind of spurious failures.

Anyway, if it is decided to change the KLSM, I think is better that you @schuay do it since you are more familiar with the KLSM code. I will go ahead and do the other changes that you have suggested so we can finish this pull request.

schuay replied:
That paper was published before this standalone version existed, and describes the implementation in the task scheduling framework Pheet (https://launchpad.net/pheet), which also evolved over time. Implementation differences are expected.

Anyway, I agree that we should proceed with landing this PR soon.

schuay reviewed (commented):
Thanks, looks almost ready to land. I have some final comments.

kjellwinblad commented:
Just made the changes that you suggested to the multi queue. Thanks for the feedback. I think it should be ready to merge now unless you find anything else.

schuay commented:
Thanks, great to have this!
